tests/models/test_validator.py

```diff
@@ -0,0 +1,112 @@
+"""Tests for the event validator"""
+
+import json
+import logging
+
+import pytest
+
+from ralph.exceptions import BadFormatException, UnknownEventException
+from ralph.models.selector import ModelSelector
+from ralph.models.validator import Validator
+from ralph.utils import get_dict_value_from_path
+
+from tests.fixtures.edx.browser import PageCloseBrowserEventFactory
+from tests.fixtures.edx.server import ServerEventFactory
+
+
+@pytest.mark.parametrize(
+    "events,errors,exceptions",
+    [
+        # No events implies no error messages nor exceptions.
+        ([], [], {}),
+        # When events are not JSON strings we should write an error message.
+        # And if ignore_errors is set to False we should raise a BadFormatException.
+        (
+            ["", 1, None, {}],
+            ["Input event is not a valid JSON string"] * 4,
+            {"ignore_errors": {False: BadFormatException}},
+        ),
+        # When events don't match any pydantic model we should write an error message.
+        # And if fail_on_unknown is set to True we should raise a UnknownEventException.
+        (
+            [
+                json.dumps({}),
+                json.dumps({"event_source": "browser"}),
+                json.dumps({"event_source": "browser", "event_type": None}),
+            ],
+            ["No matching pydantic model found for input event"] * 3,
+            {"fail_on_unknown": {True: UnknownEventException}},
+        ),
+        # When events match a pydantic model but the model validation fails
+        # we should write an error message.
+        # And if ignore_errors is set to False we should raise a BadFormatException.
+        (
+            [json.dumps({"event_source": "browser", "event_type": "page_close"})],
+            ["Input event is not a valid PageCloseBrowserEventModel"],
+            {"ignore_errors": {False: BadFormatException}},
+        ),
+    ],
+)
+@pytest.mark.parametrize("ignore_errors", [True, False])
+@pytest.mark.parametrize("fail_on_unknown", [True, False])
+def test_models_validator_validate_with_invalid_events(
+    events,
+    errors,
+    exceptions,
+    ignore_errors,
+    fail_on_unknown,
+    caplog,
+):  # pylint: disable=too-many-arguments
+    """Given a list of invalid values the validate method should log the errors and not yield."""
+
+    result = Validator(ModelSelector()).validate(events, ignore_errors, fail_on_unknown)
+    exception = get_dict_value_from_path(exceptions, ["ignore_errors", ignore_errors])
+    if not exception:
+        exception = get_dict_value_from_path(
+            exceptions, ["fail_on_unknown", fail_on_unknown]
+        )
+    with caplog.at_level(logging.ERROR):
```

Comment:
@SergioSim I don't see clearly the point of using this context in the test. Could you help me to understand how you designed this test? :)

Reply:
Good point) Thanks) Indeed, the way the test was written doesn't made it easy to read which is something to avoid. I tried to simplify now.
The `validate` method yields a generator. It will not execute until we ask for. When calling `list(result)` we ask it to run and put its results in a list until it has no more values to yield. It's at this point where logs will be written or/and exceptions raised.